# Import libraries

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
#from tabulate import tabulate
#c%config InlineBackend.figure_format = 'retina'

# Load dataset

In [11]:
df = pd.read_excel("diem_f2_md.xlsx")

In [43]:
print(df.NHHK.unique())
print(df.NAMHOC.unique())
print(df.TENHK.unique())


[162 163 171 172  71  72]
[2016 2017 2007]
[2 3 1]


In [13]:
df.head(20)

NAMHOC  TENHK  NHHK  F_MAMH              F_TENMHVN  F_DVHT  F_PHTRAMKT  \
0     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
1     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
2     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
3     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
4     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
5     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
6     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
7     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
8     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
9     2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
10    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
11    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
12    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
13    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
14    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
15    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
16    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
17    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
18    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   
19    2016      2   162  CO1003  Nhập môn về lập trình     3.0           0   

    F_PHTRAMTH F_MANH F_TO  ...    TN TILETN   THI TILETHI TKET F_DIEM1  \
0          100    L06    A  ...  10.0   30.0   5.5    40.0  8.0     NaN   
1          100    L06    A  ...   9.0   30.0   3.5    40.0  6.5     NaN   
2          100    L06    A  ...  10.0   30.0   5.5    40.0  8.0     NaN   
3          100    L06    A  ...  10.0   30.0   2.0    40.0  2.0     NaN   
4          100    L06    A  ...   6.5   30.0   2.5    40.0  2.5     NaN   
5          100    L06    A  ...   9.5   30.0   5.5    40.0  7.5     NaN   
6          100    L06    A  ...   9.0   30.0   3.0    40.0  6.0     NaN   
7          100    L06    A  ...   9.0   30.0   2.0    40.0  2.0     NaN   
8          100    L06    A  ...  10.0   30.0   7.5    40.0  8.5     NaN   
9          100    L06    A  ...  10.0   30.0   7.5    40.0  8.5     NaN   
10         100    L06    A  ...   9.5   30.0   2.5    40.0  2.5     NaN   
11         100    L06    A  ...  10.0   30.0   3.5    40.0  7.0     NaN   
12         100    L06    A  ...  10.0   30.0   8.0    40.0  9.0     NaN   
13         100    L06    A  ...  10.0   30.0   3.5    40.0  6.5     NaN   
14         100    L06    A  ...  10.0   30.0   3.5    40.0  7.0     NaN   
15         100    L06    A  ...   9.0   30.0  10.0    40.0  9.5     NaN   
16         100    L06    A  ...   4.5   30.0   6.5    40.0  6.5     NaN   
17         100    L06    A  ...   9.0   30.0   6.5    40.0  8.0     NaN   
18         100    L06    A  ...   9.0   30.0   7.0    40.0  8.0     NaN   
19         100    L06    B  ...  10.0   30.0   7.5    40.0  9.0     NaN   

    F_DIEM2  F_DIEM10  GHICHU   T  
0       NaN       NaN     NaN NaN  
1       NaN       NaN     NaN NaN  
2       NaN       NaN     NaN NaN  
3       NaN       NaN     NaN NaN  
4       NaN       NaN     NaN NaN  
5       NaN       NaN     NaN NaN  
6       NaN       NaN     NaN NaN  
7       NaN       NaN     NaN NaN  
8       NaN       NaN     NaN NaN  
9       NaN       NaN     NaN NaN  
10      NaN       NaN     NaN NaN  
11      NaN       NaN     NaN NaN  
12      NaN       NaN     NaN NaN  
13      NaN       NaN     NaN NaN  
14      NaN       NaN     NaN NaN  
15      NaN       NaN     NaN NaN  
16      NaN       NaN     NaN NaN  
17      NaN       NaN     NaN NaN  
18      NaN       NaN     NaN NaN  
19      NaN       NaN     NaN NaN  

[20 rows x 33 columns]

In [7]:
!ls

diem_f1_md.xlsx  diem_f4_md.xlsx  diem_f7_md.xlsx
diem_f2_md.xlsx  diem_f5_md.xlsx  Preprocess-EDA-1.ipynb
diem_f3_md.xlsx  diem_f6_md.xlsx  test.ipynb


# Data cleaning

In [44]:
print(df.shape)
print(df.columns)
print(df.info())

(477953, 33)
Index(['NAMHOC', 'TENHK', 'NHHK', 'F_MAMH', 'F_TENMHVN', 'F_DVHT',
       'F_PHTRAMKT', 'F_PHTRAMTH', 'F_MANH', 'F_TO', 'STT', 'F_MAKH',
       'F_TENLOP', 'F_KHOI', 'F_MANG', 'F_TENNGVN', 'MASV1', 'KT', 'TILEKT',
       'BT', 'TILEBT', 'BTLDA', 'TILEBTLDA', 'TN', 'TILETN', 'THI', 'TILETHI',
       'TKET', 'F_DIEM1', 'F_DIEM2', 'F_DIEM10', 'GHICHU', 'T'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
Int64Index: 477953 entries, 0 to 499999
Data columns (total 33 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   NAMHOC      477953 non-null  int64  
 1   TENHK       477953 non-null  int64  
 2   NHHK        477953 non-null  int64  
 3   F_MAMH      477953 non-null  object 
 4   F_TENMHVN   477953 non-null  object 
 5   F_DVHT      477953 non-null  float64
 6   F_PHTRAMKT  477953 non-null  int64  
 7   F_PHTRAMTH  477953 non-null  int64  
 8   F_MANH      477953 non-null  object 
 9   F_TO        477953 non-null  

In [50]:
df.loc[(df.F_DIEM2.notnull()) & (df.F_DIEM10.isnull())]

NAMHOC  TENHK  NHHK  F_MAMH                 F_TENMHVN  F_DVHT  \
274337    2007      1    71  806013  Hhọa vẽ kỹ thuật 1 (ckt)     2.0   
288710    2007      1    71  210015              Truyền nhiệt     2.0   
295148    2007      1    71  405024            Mạch điện tử 2     3.0   
317116    2007      1    71  405026       Hệ thống viễn thông     3.0   
337267    2007      1    71  801040         Thiết kế cầu thép     3.0   
339151    2007      1    71  602011             Hóa phân tích     2.0   
346877    2007      1    71  402701  Đtử học tương tự&ứngdụng     3.0   
349747    2007      1    71  701081   Mô phỏng trg kinh doanh     2.0   
354071    2007      1    71  701063      Quản lý doanh nghiệp     3.0   
355360    2007      1    71  003103                 Anh văn 3     2.0   
355375    2007      1    71  602011             Hóa phân tích     2.0   
363514    2007      1    71  008001    Pháp luật Việt Nam Đ/c     2.0   
363515    2007      1    71  604045             Hóa đại cương     3.0   
373343    2007      1    71  006038                    Toán 1     2.0   
373346    2007      1    71  007014                  Vật lý 1     3.0   
390245    2007      1    71  804105  Kết cấu bêtông2- gạch đá     3.0   
405242    2007      1    71  210014     Nhiệt động lực học Kt     2.0   
422964    2007      2    72  402024          Kỹ thuật điện tử     2.0   
466090    2007      2    72  805031                  Nền móng     2.0   
482813    2007      2    72  404019      Đo lường công nghiệp     2.0   

        F_PHTRAMKT  F_PHTRAMTH F_MANH F_TO  ...  TN TILETN THI TILETHI  TKET  \
274337          40          60      1    B  ... NaN    NaN NaN     NaN   4.0   
288710          20          80      2    A  ... NaN    NaN NaN     NaN   3.0   
295148          20          80      3    B  ... NaN    NaN NaN     NaN   3.5   
317116           0         100      2    A  ... NaN    NaN NaN     NaN   5.0   
337267          30          70      1    A  ... NaN    NaN NaN     NaN   5.0   
339151          30          70      4    C  ... NaN    NaN NaN     NaN   2.0   
346877          20          80      1    B  ... NaN    NaN NaN     NaN  10.0   
349747          50          50   Q201    A  ... NaN    NaN NaN     NaN   5.0   
354071          40          60      5    C  ... NaN    NaN NaN     NaN   4.0   
355360          20          80      5    B  ... NaN    NaN NaN     NaN   7.0   
355375          30          70      3    A  ... NaN    NaN NaN     NaN   5.0   
363514          20          80    L15    B  ... NaN    NaN NaN     NaN   7.0   
363515          30          70    L07    B  ... NaN    NaN NaN     NaN   4.5   
373343          20          80    L12    B  ... NaN    NaN NaN     NaN   2.5   
373346          20          80    L15    B  ... NaN    NaN NaN     NaN   7.0   
390245          40          60      6    A  ... NaN    NaN NaN     NaN   5.0   
405242          20          80   DT03    B  ... NaN    NaN NaN     NaN   8.5   
422964          20          80     11    B  ... NaN    NaN NaN     NaN   5.0   
466090          20          80      4    B  ... NaN    NaN NaN     NaN   4.0   
482813          20          80      1    A  ... NaN    NaN NaN     NaN   4.0   

       F_DIEM1  F_DIEM2  F_DIEM10  GHICHU   T  
274337    14.0      4.0       NaN     NaN NaN  
288710    14.0      3.0       NaN     NaN NaN  
295148    14.0      3.5       NaN     NaN NaN  
317116    14.0      5.0       NaN     NaN NaN  
337267    14.0      5.0       NaN     NaN NaN  
339151    14.0      2.0       NaN     NaN NaN  
346877    14.0     10.0       NaN     NaN NaN  
349747    14.0      5.0       NaN     NaN NaN  
354071    14.0      4.0       NaN     NaN NaN  
355360    14.0      7.0       NaN     NaN NaN  
355375    14.0      5.0       NaN     NaN NaN  
363514    14.0      7.0       NaN     NaN NaN  
363515    14.0      4.5       NaN     NaN NaN  
373343    14.0      2.5       NaN     NaN NaN  
373346    14.0      7.0       NaN     NaN NaN  
390245    14.0      5.0  

In [47]:
df.loc[(df.F_DIEM10 == df.TKET) & (df.F_DIEM10.notnull())]

NAMHOC  TENHK  NHHK  F_MAMH         F_TENMHVN  F_DVHT  F_PHTRAMKT  \
260732    2007      1    71  001011     Cnxh khoa học     2.0          30   
260733    2007      1    71  001011     Cnxh khoa học     2.0          30   
260734    2007      1    71  001011     Cnxh khoa học     2.0          30   
260735    2007      1    71  001011     Cnxh khoa học     2.0          30   
260736    2007      1    71  001011     Cnxh khoa học     2.0          30   
...        ...    ...   ...     ...               ...     ...         ...   
499995    2007      2    72  809031  Sức bền vật liệu     4.0           0   
499996    2007      2    72  809031  Sức bền vật liệu     4.0           0   
499997    2007      2    72  809031  Sức bền vật liệu     4.0           0   
499998    2007      2    72  809031  Sức bền vật liệu     4.0           0   
499999    2007      2    72  809031  Sức bền vật liệu     4.0           0   

        F_PHTRAMTH F_MANH F_TO  ...  TN TILETN THI TILETHI TKET F_DIEM1  \
260732          70      2    A  ... NaN    NaN NaN     NaN  7.1     5.0   
260733          70      2    A  ... NaN    NaN NaN     NaN  7.2     6.5   
260734          70      2    A  ... NaN    NaN NaN     NaN  8.1     7.0   
260735          70      2    A  ... NaN    NaN NaN     NaN  7.9     6.5   
260736          70      2    A  ... NaN    NaN NaN     NaN  5.4     4.0   
...            ...    ...  ...  ...  ..    ...  ..     ...  ...     ...   
499995         100      1    D  ... NaN    NaN NaN     NaN  1.0     NaN   
499996         100      1    D  ... NaN    NaN NaN     NaN  3.0     NaN   
499997         100      1    D  ... NaN    NaN NaN     NaN  5.0     NaN   
499998         100      1    D  ... NaN    NaN NaN     NaN  3.0     NaN   
499999         100      1    D  ... NaN    NaN NaN     NaN  0.0     NaN   

        F_DIEM2  F_DIEM10  GHICHU   T  
260732      8.0       7.1     NaN NaN  
260733      7.5       7.2     NaN NaN  
260734      8.5       8.1     NaN NaN  
260735      8.5       7.9     NaN NaN  
260736      6.0       5.4     NaN NaN  
...         ...       ...     ...  ..  
499995      1.0       1.0     NaN NaN  
499996      3.0       3.0     NaN NaN  
499997      5.0       5.0     NaN NaN  
499998      3.0       3.0     NaN NaN  
499999     13.0       0.0     NaN NaN  

[237787 rows x 33 columns]

In [29]:
df.iloc[498520]

NAMHOC                       2007
TENHK                           2
NHHK                           72
F_MAMH                     501710
F_TENMHVN           Mạng máy tính
F_DVHT                        3.0
F_PHTRAMKT                     20
F_PHTRAMTH                     80
F_MANH                          1
F_TO                            B
STT                           NaN
F_MAKH                         VP
F_TENLOP                   VP04HK
F_KHOI                  DHVPVHK04
F_MANG                        VHK
F_TENNGVN     Kỹ thuật Hàng không
MASV1                     9365812
KT                            NaN
TILEKT                        NaN
BT                            NaN
TILEBT                        NaN
BTLDA                         NaN
TILEBTLDA                     NaN
TN                            NaN
TILETN                        NaN
THI                           NaN
TILETHI                       NaN
TKET                          8.5
F_DIEM1                       6.5
F_DIEM2       

0. Names of some columns have space => not ideal for future operations. The additional part after space is only for clarification => also unnecessary. We can safely remove those part in the names.

**NOT NEEDED ANYMORE**

In [ ]:
"""
df.rename(columns = 
          {
           "F_MAMH (Mã Môn Học)": "F_MAMH",
           "F_TENMHVN (Tên Môn Học)": "F_TENMHVN",
           "F_MANH (Mã Nhóm Học?)": "F_MANH",
           "KT (Thi giữa kì)": "KT",
           "TILEKT (% thi giữa kì chiếm trong điểm tổng kết)":"TILEKT",
           "BT (Bài tập trên lớp)":"BT",
           "TILEBT (%)":"TILEBT",
           "BTLDA (Bài tập lớn)":"BTLDA",
           "TN (Điểm lab)":"TN",
           "THI (điểm thi)":"THI",
           "TKET (điểm tổng kết)":"TKET"
          }, inplace = True)
"""

1. Drop duplicates

In [9]:
df.drop_duplicates(keep = "first", inplace = True)

2. Column F_DIEM10 and T doesn't have any non-null value => Should be dropped

In [9]:
df.drop(["F_DIEM10", "T"], axis = 1, inplace = True)

3. Columns NAMHOC and TENHK are redundant with the existence of the column NHHK (2015 + HK1 -> 151). They will be dropped



In [10]:
df.drop(["NAMHOC", "TENHK"], axis = 1, inplace = True)

4. Điểm tổng kết should be of type float64, not string. The reason that it is currenly string is because some special grades got mixed up as well ('CT', 'VT', 'RT', 'IH','HT', 'KD', 'DT' and numeric >10 scores)

![image info](https://raw.githubusercontent.com/qdgiang/stuffs/main/Screenshot%202022-04-05%20225225.png)

As we can see, the special scores are for special cases only, therefore we will drop them all since special predictions like being barred from taking an exam for score **CT** and being absent from an exam for score **VT** require additional social information which we do not currently possess.




In [30]:
df.TKET.unique()

array([ 8.  ,  6.5 ,  2.  ,  2.5 ,  7.5 ,  6.  ,  8.5 ,  7.  ,  9.  ,
        9.5 ,  1.  ,  0.  ,  1.5 , 13.  ,  4.  ,  5.5 ,  5.  ,  4.5 ,
        3.5 ,  3.  , 10.  ,  0.5 ,  8.11,  7.71,  9.16,  8.88,  9.36,
        8.83,  8.33,  8.44,  9.44,  9.38, 11.  , 15.  , 22.  ,  8.7 ,
        8.6 ,  8.9 ,  7.2 ,  8.4 ,  7.7 ,  7.1 ,  9.2 ,  8.8 ,  8.57,
        7.64,  8.64,  7.41,  9.33,  7.58,  7.86,  9.1 ,  8.3 ,  9.4 ,
        8.1 ,  9.8 ,  7.92,  6.58,  7.67,  8.08,  6.75,  7.83,  6.4 ,
        7.75,  7.8 ,  6.83,  8.2 ,  7.9 ,  7.3 ,  9.35,  9.3 ,  6.7 ,
        9.15,  7.6 ,  9.6 ,  8.75,  8.95,  9.05,  9.25,  7.4 ,  7.93,
        7.78,  8.43,  7.79,  8.67,  8.92,  7.14,  8.86,  9.71,  8.58,
        8.07,  6.71,  8.29,  8.42,  9.17,  7.25,  9.21,  7.57,  8.41,
        8.79,  8.91,  9.14,  7.43,  7.33,  9.75,  7.36,  8.17,  8.36,
        7.17,  9.29,  8.46,  8.93,  6.6 ,  6.25,  9.43,  8.87,  9.28,
        8.21,  8.14,  8.71,  6.93,  5.79,  8.78,  7.21,  6.86,  7.56,
        9.52,  8.97,

In [32]:
special_score = ['CT', 'MT', 'VT', 'RT', 'IH', 'HT', 'KD', 'DT', 'VP', 11, 12, 13, 14, 15, 16, 17, 20, 21, 22]

In [33]:
df = df.query('TKET not in @special_score')

In [34]:
df.shape

(477953, 33)

In [35]:
df.TKET = pd.to_numeric(df.TKET)

/tmp/ipykernel_30592/2744034334.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.TKET = pd.to_numeric(df.TKET)


5. After deleting all rows with special final scores, the column GHICHU is now full of empty values, suggesting that this column is for clarifying special scores. Now that we have done step 4, there's no need for this column any more.

In [17]:
df.drop(["GHICHU"], axis = 1, inplace = True)

6. A **subject** can sometimes have multiple **subject codes**. This can be explained by the fact that the university has 2 parallel system, normal courses using **Vietnamese** and high quality courses using **English**

Moreover, the consistency of the subject name is also not good: "Những ngLý cơbản cn M-Lê" and "Những NgLý cơbản CN M-Lê" should be one subject, but they are not in this dataset.

Therefore, for consistency, we will use subject code column in this dataset and get the subject name from official university source.

In [1]:
print(len(df["F_TENMHVN"].unique()))
print(len(df["F_MAMH"].unique()))
print("Giải tích 1 has 3 subject codes: "+ str(df[df['F_TENMHVN'] == "Giải tích 1"]["F_MAMH"].unique()))
print("Những ngLý cơbản cn M-Lê has the subject code: "+ str(df[df['F_TENMHVN'] == "Những ngLý cơbản cn M-Lê"]["F_MAMH"].unique()))
print("Những NgLý cơbản CN M-Lê has the subject code: "+ str(df[df['F_TENMHVN'] == "Những NgLý cơbản CN M-Lê"]["F_MAMH"].unique()))

NameError: name 'df' is not defined

In [36]:
df.F_MAMH = df.F_MAMH.astype(str)

/tmp/ipykernel_30592/1348358720.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.F_MAMH = df.F_MAMH.astype(str)


In [37]:
print(df.F_MAMH.unique()[:20])

['CO1003' 'SP1005' '805012' 'CI2029' 'MA2017' '215122' '402706' '405103'
 'EE3151' 'PE1005' '260601' '400400' '402708' '405102' '405414' 'EE3117'
 '213066' '260619' '409036' '700213']


We can see that some subject codes only have 4 digits, which is not true for our university. This is because Excel automatically convert those codes into int, which promptly eliminated the first two 00 digits.

A quick verification on the web site confirms that the codes 006001 and 006401 do correspond to the subject Calculus 1, and the same for 001001 to Marxist - Leninist Philosophy

In [40]:
print(df[df.F_MAMH.str.len() > 6].F_MAMH.unique())
print(df[df.F_MAMH.str.len() < 6].F_MAMH.unique())

[]
[]


In [39]:
df.loc[df.F_MAMH.str.len() == 4, 'F_MAMH'] = "00" + df.F_MAMH

7. By checking with df.info() again, we can see that some columns still have Null values left.

Column F_TO is for subclass of each class (CC01_A, CC05_D, ..). This information is unnecessary in building models, but could be useful during visualization process as we compare performance between different classes => Column won't be drop


---


Closer inspection reveals that some columns have the same number of non-null: KT and TILEKT, BT and TILEBT, BTLDA and TILEBTLDA, THI and TILETHI. This is evidently because when a subject doesn't have midterm, for example, then both the score and the % of the midterm won't be put into the dataset.

We double check to confirm our theory

In [41]:
df.F_TO.unique()

array(['A', 'B', '_', 'C', 'D', 'E', 'AA', 'AB', 'BA', 'BB', 'AC', 'AD',
       'F', 'G', 'I', 'H', 'J', 'Q'], dtype=object)

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 437953 entries, 0 to 499999
Data columns (total 28 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   NHHK        437953 non-null  int64  
 1   F_MAMH      437953 non-null  object 
 2   F_TENMHVN   437953 non-null  object 
 3   F_DVHT      437953 non-null  float64
 4   F_PHTRAMKT  437953 non-null  int64  
 5   F_PHTRAMTH  437953 non-null  int64  
 6   F_MANH      437953 non-null  object 
 7   F_TO        352678 non-null  object 
 8   STT         384131 non-null  float64
 9   F_MAKH      437953 non-null  object 
 10  F_TENLOP    437953 non-null  object 
 11  F_KHOI      437953 non-null  object 
 12  F_MANG      437953 non-null  object 
 13  F_TENNGVN   437953 non-null  object 
 14  MASV1       437953 non-null  int64  
 15  KT          360583 non-null  float64
 16  TILEKT      360583 non-null  float64
 17  BT          319599 non-null  float64
 18  TILEBT      319599 non-null  float64
 19  BT

In [42]:
df.shape

(477953, 33)

In [40]:
print(df[df.KT.isnull() & df.TILEKT.isnull()].shape[0])
print(437953 - 360583) # We confirm that everytime a KT is null, TILEKT is null as well, and vice versa
print(df[df.BT.isnull() & df.TILEBT.isnull()].shape[0])
print(437953 - 319599) # We confirm that everytime a BT value is null, TILEBT is null as well, and vice versa
print(df[df.BTLDA.isnull() & df.TILEBTLDA.isnull()].shape[0])
print(437953 - 318543) # We confirm that everytime a BTLDA value is null, TILEBTLDA is null as well, and vice versa
print(df[df.THI.isnull() & df.TILETHI.isnull()].shape[0])
print(437953 - 419888) # We confirm that everytime a THI value is null, TILETHI is null as well, and vice versa

77370
77370
118354
118354
119410
119410
18065
18065


Here we encounter an additional problem: Some students only have a TKET score without any component score, likely due to mistake while inputting data. We can see an example below

In [41]:
df[df.THI.isnull() & df.TILETHI.isnull()].iloc[1]

NHHK                               151
F_MAMH                          001001
F_TENMHVN     Những ngLý cơbản cn M-Lê
F_DVHT                             5.0
F_PHTRAMKT                          40
F_PHTRAMTH                          60
F_MANH                            TT01
F_TO                                 A
STT                                8.0
F_MAKH                              PD
F_TENLOP                      CT14TIE2
F_KHOI                       DHTTIEN14
F_MANG                             IEN
F_TENNGVN          Hệ thống Năng lượng
MASV1                          6070207
KT                                 NaN
TILEKT                             NaN
BT                                 NaN
TILEBT                             NaN
BTLDA                              NaN
TILEBTLDA                          NaN
TN                                 NaN
TILETN                             NaN
THI                                NaN
TILETHI                            NaN
TKET                     

=> Solution: When building models that use component scores, these rows will be dropped. Else when only using final score, the missing component scores won't be used, meaning there will be no problem at all.

An outlier is the case with TN and TILETN. We have 256622 non-null TN values, but 295466 non-null TILETN values.

In [ ]:
print(df.TN.count())
print(df.TILETN.count())
print(df[df.TN.isnull() & df.TILETN.notnull()].shape[0])
print(df.TILETN.count() - df.TN.count()) # We confirm that the difference between TN and TILETN columns is solely due to TILETN being non-null in some places where TN is null 

256622
295466
38844
38844


In [54]:
print(df[df.TN.isnull()].TILETN.value_counts())
print(df[df.TN.isnull()].TILETN.unique())

0.0    38844
Name: TILETN, dtype: int64
[nan  0.]


Example

In [58]:
df[df.TN.isnull() & df.TILETN.notnull()].iloc[1]

NHHK                             162
F_MAMH                        001004
F_TENMHVN     Đ/lối c/mạng đảng csvn
F_DVHT                           3.0
F_PHTRAMKT                        30
F_PHTRAMTH                        70
F_MANH                           A03
F_TO                               A
STT                              2.0
F_MAKH                            CK
F_TENLOP                    CK12CK01
F_KHOI                     DHCKCKI12
F_MANG                           CKI
F_TENNGVN          Cơ khí-Cơ điện tử
MASV1                       29874580
KT                               3.2
TILEKT                          30.0
BT                               0.0
TILEBT                           0.0
BTLDA                            0.0
TILEBTLDA                        0.0
TN                               NaN
TILETN                           0.0
THI                              3.5
TILETHI                         70.0
TKET                             3.5
F_DIEM1                          NaN
F

With this, it's easy to see that the NULL values in all these column is simply because subjects don't require all component score, but only some. Therefore, no drastic change is required

In [ ]:
#df.TILEKT = df.TILEKT.fillna(0)
#df.TILEBT = df.TILEBT.fillna(0)
#df.TILEBTLDA = df.TILEBTLDA.fillna(0)
#df.TILETN = df.TILETN.fillna(0)
#df.TILETHI = df.TILETHI.fillna(0)

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 437953 entries, 0 to 499999
Data columns (total 28 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   NHHK        437953 non-null  int64  
 1   F_MAMH      437953 non-null  object 
 2   F_TENMHVN   437953 non-null  object 
 3   F_DVHT      437953 non-null  float64
 4   F_PHTRAMKT  437953 non-null  int64  
 5   F_PHTRAMTH  437953 non-null  int64  
 6   F_MANH      437953 non-null  object 
 7   F_TO        352678 non-null  object 
 8   STT         384131 non-null  float64
 9   F_MAKH      437953 non-null  object 
 10  F_TENLOP    437953 non-null  object 
 11  F_KHOI      437953 non-null  object 
 12  F_MANG      437953 non-null  object 
 13  F_TENNGVN   437953 non-null  object 
 14  MASV1       437953 non-null  int64  
 15  KT          360583 non-null  float64
 16  TILEKT      360583 non-null  float64
 17  BT          319599 non-null  float64
 18  TILEBT      319599 non-null  float64
 19  BT

When F_DIEM1 is null and F_DIEM2 is not null, almost all cases is F_DIEM2 = TKET

In [63]:
#print(df.loc[df.F_DIEM2.notnull()].shape)
#df.loc[(df.F_DIEM2 != df.TKET) & (df.F_DIEM2.notnull())]
df.loc[(df.F_DIEM2 != df.TKET) & (df.F_DIEM2.notnull()) & (df.F_DIEM1.isnull()) & (df.F_DIEM2 != 13)]

NHHK  F_MAMH            F_TENMHVN  F_DVHT  F_PHTRAMKT  F_PHTRAMTH  \
252733   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252734   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252737   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252739   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252740   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252741   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252742   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252744   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252745   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252748   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252753   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252755   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252756   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252759   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252761   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252765   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252770   153  610302  Thực tập tốt nghiệp     3.0           0         100   
252781   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252783   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252792   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252793   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252796   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252799   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252800   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252801   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252803   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252811   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252813   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252819   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252824   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252825   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252828   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252830   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252831   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252833   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252834   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252835   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252837   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252840   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252843   153  611302  Thực tập tốt nghiệp     3.0           0         100   
252845   153  611302  Thực tập tốt nghiệp     3.0           0         100   

       F_MANH F_TO  STT F_MAKH  ... TILEBT BTLDA TILEBTLDA  TN  TILETN  THI  \
252733    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252734    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252737    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252739    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252740    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252741    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252742    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252744    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN     NaN  NaN   
252745    A01    _  NaN     MO  ...    NaN   NaN       NaN NaN

=> In case F_DIEM2 != TKET, it's actually F_DIEM2 rounding up to become the final TKET

In case where F_DIEM1 and F_DIEM2 are both present, TKET is the average of the two scores.

Given the small amount of occurences of both F_DIEM1 and F_DIEM2, and the fact that they are, in all cases, represented in the final TKET, We can safely remove the F_DIEM1 and F_DIEM2 columns.

In [64]:
#df.drop(["F_DIEM1", "F_DIEM2"], axis = 1, inplace = True)

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 437953 entries, 0 to 499999
Data columns (total 26 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   NHHK        437953 non-null  int64  
 1   F_MAMH      437953 non-null  object 
 2   F_TENMHVN   437953 non-null  object 
 3   F_DVHT      437953 non-null  float64
 4   F_PHTRAMKT  437953 non-null  int64  
 5   F_PHTRAMTH  437953 non-null  int64  
 6   F_MANH      437953 non-null  object 
 7   F_TO        352678 non-null  object 
 8   STT         384131 non-null  float64
 9   F_MAKH      437953 non-null  object 
 10  F_TENLOP    437953 non-null  object 
 11  F_KHOI      437953 non-null  object 
 12  F_MANG      437953 non-null  object 
 13  F_TENNGVN   437953 non-null  object 
 14  MASV1       437953 non-null  int64  
 15  KT          360583 non-null  float64
 16  TILEKT      360583 non-null  float64
 17  BT          319599 non-null  float64
 18  TILEBT      319599 non-null  float64
 19  BT

In [66]:
df.to_csv('diem1_cleaned.csv', index=False)

# Exploratory data analysis

The most important thing to keep in mind is that a student's course varies across different department. A CS student and a ME student won't have the same schedule. 

In [ ]:
df.F_TENNGVN.unique()

array(['Cơ khí-Cơ điện tử', 'Cơ điện tử', 'Kỹ thuật Chế tạo',
       'KT Hệ Thống CN', 'MT&CôngNghệ ThôngTin', 'Kỹ thuật Dệt May',
       'Điện - Điện tử', 'Bảo dưỡng CN', 'Kỹ thuật Máy tính',
       'CN Hóa học &ThựcPhẩm', 'Kỹ thuật MáyTính(TN)',
       'Quản Lý Công Nghiệp', 'Cơ KT- Vật lý KT', 'Công nghệ vật liệu',
       'Kiến Trúc', 'Kỹ thuật Giao thông', 'Xây dựng', 'Cơ Điện Tử (TN)',
       'KTMáy XD &NângChuyển', 'KT ĐiệnTử -ViễnThông',
       'KT Trắc địa - Bản đồ', 'Vật liệu Xây dựng', 'Kỹ thuật Hàng không',
       'Khoa học Máy tính', 'Điện năng', 'Khoa học MáyTính(TN)',
       'ĐKhiển Tự Động (TN)', 'Điện Tử -VThông (TN)', 'Môi Trường',
       'QT-Thiết Bị & ĐKhiển', 'CôngNghệ KỹThuật ÔTô',
       'Kỹ thuật Thiết kế', 'Hệ thống Năng lượng', 'CN May - Thời trang',
       'KT Nhiệt lạnh', 'Kỹ thuật Hóa Lý', 'Công nghệ Sinh học',
       'Kỹ thuật tàu thủy', 'Kỹ thuật Hóa Hữu cơ', 'KT Hóa Vô cơ (TN)',
       'Địa chất-Dầu khí', 'Công nghệ Thực phẩm', 'KT Vật Liệu Kim Loại',
   

In [ ]:
counts = df.F_TENNGVN.value_counts()
counts.to_csv('TenNganh.csv',index=True)

As we can see, there are 3 CS major in the list: 'Khoa học Máy tính', 'Khoa học MáyTính(TN)' - likely being Tài Năng - and 'Khoa học MáyTính(B2)' - likely being Văn Bằng 2.

As Điện - Điện tử has the largest amount, we will work on this major first. Process for others should be the same.

In [ ]:
df_e = df[df.F_TENNGVN == "Điện - Điện tử"]
df_e.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68835 entries, 41 to 499987
Data columns (total 30 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   NHHK        68835 non-null  int64  
 1   F_MAMH      68835 non-null  object 
 2   F_TENMHVN   68835 non-null  object 
 3   F_DVHT      68835 non-null  float64
 4   F_PHTRAMKT  68835 non-null  int64  
 5   F_PHTRAMTH  68835 non-null  int64  
 6   F_MANH      68835 non-null  object 
 7   F_TO        55042 non-null  object 
 8   STT         58572 non-null  float64
 9   F_MAKH      68835 non-null  object 
 10  F_TENLOP    68835 non-null  object 
 11  F_KHOI      68835 non-null  object 
 12  F_MANG      68835 non-null  object 
 13  F_TENNGVN   68835 non-null  object 
 14  MASV        68835 non-null  int64  
 15  F_HOLOTVN   68835 non-null  object 
 16  F_TENVN     68835 non-null  object 
 17  KT          50671 non-null  float64
 18  TILEKT      68835 non-null  float64
 19  BT          52261 non-n

In [ ]:
df_e.MASV.value_counts()

76999866    49
30925388    48
27984551    48
75127117    48
57325308    47
            ..
75928204     1
19689013     1
77011049     1
92772795     1
13426086     1
Name: MASV, Length: 3346, dtype: int64

In [ ]:
df_student = df_e[df_e.MASV == 7287845] # Working on 1 student

In [ ]:
df_student.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46 entries, 438 to 487708
Data columns (total 30 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   NHHK        46 non-null     int64  
 1   F_MAMH      46 non-null     object 
 2   F_TENMHVN   46 non-null     object 
 3   F_DVHT      46 non-null     float64
 4   F_PHTRAMKT  46 non-null     int64  
 5   F_PHTRAMTH  46 non-null     int64  
 6   F_MANH      46 non-null     object 
 7   F_TO        37 non-null     object 
 8   STT         42 non-null     float64
 9   F_MAKH      46 non-null     object 
 10  F_TENLOP    46 non-null     object 
 11  F_KHOI      46 non-null     object 
 12  F_MANG      46 non-null     object 
 13  F_TENNGVN   46 non-null     object 
 14  MASV        46 non-null     int64  
 15  F_HOLOTVN   46 non-null     object 
 16  F_TENVN     46 non-null     object 
 17  KT          36 non-null     float64
 18  TILEKT      46 non-null     float64
 19  BT          37 non-null  

In [ ]:
df_e.F_TENMHVN.value_counts()

Thực tập điện 1           1641
Xử lý số tín hiệu         1637
Thực tập điện tử 1        1615
Mạch điện tử              1542
Thực tập điện tử 2        1492
                          ... 
Xử lý tín hiệu số            1
Lập và phân tích dự án       1
Kiến trúc máy tính           1
Quản lý sản xuất             1
Anh văn thương mại 2         1
Name: F_TENMHVN, Length: 223, dtype: int64

In [ ]:
def draw_his_1_subject(subject_name: str, df):
  fig = px.histogram(df[df.F_TENMHVN == subject_name], x= "TKET")
  fig.show()


In [ ]:
draw_his_1_subject("Mạch điện tử", df_e) # See the score distribution of one particular subject

In [ ]:
def draw_his_1_student(student_ID: int, df):
  fig = px.bar(df[df.MASV == student_ID], y='TKET', x='F_TENMHVN', text = "NHHK", color = "NHHK")
  fig.update_layout(barmode = 'group')
  fig.update_layout(uniformtext_minsize=8, uniformtext_mode='hide')
  #fig.add_hrect(y0=0, y1=4.99999, line_width=0, fillcolor="red", opacity=0.3)
  fig.show()

In [ ]:
draw_his_1_student(7287845, df_e)

We can see in some subjects, this student doesn't achieve the required score to pass, which then lead to them trying to finish them as soon as possible. For example, for subject **Thí nghiệm Mạch điện tử** in semester 151 and **Máy điện** in 152, he both got a score of 4. This led to his decision to take both these two subjects again in the summer semester 153.

Understandably, semester 161, with 12 different classes, is too much for him to retake anymore subjects.

And for semester 162, for subject **Vi xử lý** and **Bảo vệ relay trong htd** which he failed during semester 161, he immediately retake them again in 162, as well as two subjects Kỹ thuật số and Giải tích mạch, which he failed in 151.

In short, recommending subjects that students have yet to pass is a valid suggestion, due to their desire to finish what they have started while their knowledge on the subject is still there.



---

TBC - On delay bc midterm

In [ ]:
def bruh():
  try:

    x = 16

    return x

  except Error:

    return x + 1

  else:

    return x + 2

  finally:

    return x + 3

In [ ]:
print(bruh())

19
